File: go/300.600/386.LexicographicalNumbers.go
package main

import "fmt"

func lexicalOrder(n int) []int {
	ret := make([]int, 0, n)
	var dfs func(b int)
	dfs = func(b int) {
		b *= 10
		for i := 0; i < 10; i++ {
			if b+i > n {
				break
			}
			ret = append(ret, b+i)
			dfs(b + i)
		}
	}
	for i := 1; i < 10; i++ {
		if i > n {
			break
		}
		ret = append(ret, i)
		dfs(i)
	}
	return ret
}

func main() {
	fmt.Println(lexicalOrder(13))
	fmt.Println(lexicalOrder(2))
	fmt.Println(lexicalOrder(200))
}

File: 300.500/386.LexicographicalNumbers.cpp
#include <iostream>
#include <vector>
using namespace std;

class LexicographicalNumbers{
public:
    vector<int> lexicalOrder(int n) {
        ret.clear();
        ret.resize(n, 0);
        end = n;
        pos = 0;
        for(int i = 1;i<10;++i){
            if(i>end)
                break;
            ret[pos++] = i;
            loop(i);
        }
        return ret;
    }
private:
    void loop(int b){
        b*=10;
        for(int i = 0; i<10;++i){
            if(b+i>end)
                break;
            ret[pos++]=b+i;
            loop(b+i);
        }
    }
private:
    vector<int> ret;
    int end{};
    int pos{};
};

int main(int argc, char const *argv[]){
    LexicographicalNumbers cls;
    auto& ret = cls.lexicalOrder(13);
    auto& ret1 = cls.lexicalOrder(2);
    auto& ret2 = cls.lexicalOrder(200);
    return 0;
}
